int32[] left_data
int32[] right_data
